tensorboard/data/server/downsample.rs

```diff
@@ -0,0 +1,109 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+//! Downsampling for in-memory sequences.
+
+use rand::SeedableRng;
+use rand_chacha::ChaCha20Rng;
+
+/// Downsamples `xs` in place to contain at most `k` elements, always including the last element.
+///
+/// If `k == 0`, then `xs` is cleared. If `k >= xs.len()`, then `xs` is returned unchanged.
+/// Otherwise, a uniformly random sample of `k - 1` of the first `n - 1` elements of `xs` is chosen
+/// and retained, and the final element of `xs` is also retained. The relative order of elements of
+/// `xs` is unchanged.
+///
+/// More declaratively: among all subsequences of `xs` of length `min(k, xs.len())` that include
+/// the last element, one is selected uniformly at random, and `xs` is updated in place to
+/// represent that subsequence.
+///
+/// The random number generator is initialized with a fixed seed, so this function is
+/// deterministic.
+pub fn downsample<T>(xs: &mut Vec<T>, k: usize) {
+    let n = xs.len();
+    if k == 0 {
+        xs.clear();
+        return;
+    }
+    if k >= n {
+        return;
+    }
+
+    let mut rng = ChaCha20Rng::seed_from_u64(0);
+
+    // Choose `k - 1` of the `n - 1` indices to keep, and move their elements into place. Then,
+    // move the last element into place and drop extra elements.
+    let mut indices = rand::seq::index::sample(&mut rng, n - 1, k - 1).into_vec();
+    indices.sort_unstable();
+    for (dst, src) in indices.into_iter().enumerate() {
+        xs.swap(dst, src);
+    }
+    xs.swap(k - 1, n - 1);
+    xs.truncate(k);
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    /// Clones `xs` and [`downsample`]s the result to `k` elements.
+    fn downsample_cloned<T: Clone>(xs: &Vec<T>, k: usize) -> Vec<T> {
+        let mut ys = xs.clone();
+        downsample(&mut ys, k);
+        ys
+    }
+
+    #[test]
+    fn test_deterministic() {
+        let xs: Vec<char> = "abcdefg".chars().collect();
+        let expected = downsample_cloned(&xs, 4);
+        assert_eq!(expected.len(), 4);
+        for _ in 0..100 {
+            assert_eq!(downsample_cloned(&xs, 4), expected);
+        }
+    }
+
+    #[test]
+    fn test_underlong_ok() {
```

Comment:
what is "underlong"?

Reply:
Converse of [*overlong*][1], which is somewhat common in encoding
contexts (e.g., “Modified UTF-8 is UTF-8 but with the 2-overlong
encoding of NUL”). Granted, I made up *underlong*, but it sounds better
than *overshort*. :-)

I can go with `test_k_less_than_n_ok` or something if you prefer.

[1]: https://stackoverflow.com/questions/7113117/what-is-exactly-an-overlong-form-encoding

Reply:
Yeah, I'd prefer a more straightforward name - even "overlong" would be obscure IMO given that it's not a particularly common term (as the SO question itself indicates) and its applicability to this scenario is far from obvious to me.  For that matter isn't this `test_k_greater_than_n_ok`?  So I'm not sure I even understand the metaphor despite the intention having been explained to me :)

Reply:
Ah, my bad; “*k* greater than *n*” is indeed what I meant. See, this is
why I like having a name for the concept rather than fiddling with the
algebra. ;-)

Done, as `test_ok_when_k_greater_than_n`.